packages/realm/src/app-services/App.ts

```diff
@@ -140,6 +141,13 @@ export type AppConfiguration = {
    * @since 12.2.0
    */
   metadata?: Metadata;
+
+  /**
+   * Specify a custom `fetch` implementation.
+   * @since 12.3.0
+   * @experimental This API is experimental and may change or be removed.
+   */
+  fetchOverride?: unknown;
```

Comment:
This type should match the `NetworkType` from `network.ts`
```
  fetch(request: Request | binding.Request ): Promise<FetchResponse>;
```

Reply:
I think it should be similar to a [specification](https://developer.mozilla.org/en-US/docs/Web/API/fetch). Also, I'd say we should probably try to not expose bindgen details to the public API.

Reply:
Will there be unintended side effects if the `fetch` function doesn't match our specifications?

Reply:
Are you referring to our `NetworkType`? Since our TS expects the `fetch` used (whether it'd be ours or a user-provided one) to be any of the functions defined in the `NetworkType`, the preferable option (from the way it's set up atm) seems to be `fetch(request: Request): Promise<FetchResponse>;`. In that case we'd need to export `Request` and `FetchResponse`.

But, safest would probably be to ask the user of their intended use (arg and return type) and if `fetch(request: Request): Promise<FetchResponse>;` would be sufficient for them.

Reply:
The user in this case just wants to replace `node-fetch` with `window.fetch`, since they are running `realm` in a browser environment in `electron`.  

My main concern is [this line](https://github.com/realm/realm-js/blob/main/packages/realm/src/platform/network.ts#L55).  Here there is a transformation if the request happening depending on it's contents.  If a user is providing a custom `fetch`, will not doing this transformation cause an issue?

From this observation, perhaps @kraenhansen is correct in that we should directly override the `fetch` method within `@realm/network-transport` (a level deeper), and leave this one alone.  Then we would also need to provide mechanisms to override the `AbortController`.

It's probably worth while doing a manual test in React Native, to see if such an override could cause significant issues.

Reply:
The way I understood it was that the entire [fetch](https://github.com/realm/realm-js/blob/83783da13ac4616e8955a3c35f7491986f2786ce/packages/realm/src/platform/network.ts#L53) would be replaced by the user-provided implementation. Thus the line you linked to is only for when our implementation is used, which is why the `typeof` check is necessary in that case.

Reply:
Our internal code will be potentially sending two types of requests objects to the fetch in `network.ts`.  When a user overrides this, the override may have to also react to two different types of requests.  It's possible it will just work, but I don't think we have properly tested for this.

Reply:
That could work, but thinking about that gave me another idea.  We could do something like this:

```
let baseFetch = transport.fetch;

export const overrideFetch(fetch) = {
  baseFetch = fetch;
}

export const network: NetworkType = {
  async fetch(request) {
    debug("Requesting %O", request);
    const fetchRequest = typeof request.method === "string" ? request : toFetchRequest(request);
    const response = await baseFetch();  // before was `transport.fetch()`
    debug("Responded %O", response);
    return response;
  },
};
```

Probably should do this in a way that freezes the fetch override, and maybe rename `baseFetch` to something else.

Reply:
I like it! 😎 I'll add it to our Round Table and let's discuss more :)
```ts
export function overrideFetch(fetch: (request: Request) => Promise<FetchResponse>) {
  baseFetch = fetch;
}
```

Reply:
My immediate feeling is that I like the declarative API of keeping this a configuration of the `App`. Possibly providing a way to override the default values for those config values, but then again - how many developers do we expect to have more than one `App` instance per application?

I like ☝️ over this more procedural `overrideFetch` function suggestion as that minimize the state in the library, which also has the added benefit of allowing two `App` instances to fetch using different implementations concurrently.

Reply:
The intention of the `overrideFetch` suggestion keeps the configuration through `App`, not for the developer to call this directly.  The point being, to keep the internal logic of `network:fetch` (calls to `toFetchRequest`), and give the opportunity to override the `fetch` within the `network:fetch` method.
I was not proposing to remove the configuration by `App`, but to clear up that we would be overriding the internal `network:fetch` logic with the current PR state.

Reply:
@takameyer I don't see how that'll work. The `App` would have to call `overrideFetch`, passing in the `fetch` before every single request to the server? It'll be a shared state between two `App` instances, that would have to be constantly switched between the two.

Reply:
Then we should write this in a way that works in the case you are suggesting.  The code I wrote above was only an example, but the intention should be clear.  

In any case, I think we are misunderstanding each other and would enjoy discussing this further when you return.